src/license_expression/__init__.py

```diff
@@ -617,6 +654,120 @@ def dedup(self, expression):
+
+        # Check `expression` type
+        try:
+            self.parse(expression)
+        except ExpressionError as e:
+            expression_info.errors.append(str(e))
+            return expression_info
+
+        # Check `expression` syntax
+        try:
+            self.parse(expression, strict=strict)
+        except ExpressionParseError as e:
+            expression_info.errors.append(str(e))
+            expression_info.invalid_symbols.append(e.token_string)
+            return expression_info
+
+        # Check `expression` keys
+        try:
+            parsed_expression = self.parse(expression, strict=strict, validate=True)
+        except ExpressionError as e:
+            error_message = str(e)
+            expression_info.errors.append(error_message)
+            if 'Unknown license key' in error_message:
+                unknown_keys = self.unknown_license_keys(expression)
+                expression_info.invalid_symbols.extend(unknown_keys)
+            return expression_info
+
+        # If we have not hit an exception, load `expression_info` and return it
+        symbols = list(parsed_expression.symbols)
+        expression_info.normalized_license_expression = parsed_expression.render()
+        expression_info.valid_symbols = [s.render() for s in symbols]
+        expression_info.exception_symbols = [s.render() for s in symbols if isinstance(s, LicenseWithExceptionSymbol) or s.is_exception]
+        return expression_info
+
+
+def get_license_key_info(license_key_index_location=None):
+    """
+    Return a list of dictionaries that contain license key information from
+    `license_key_index_location`
+
+    If `license_key_index_location` is not present, then we use a vendored copy
+    of the license key index from https://scancode-licensedb.aboutcode.org/
+    """
+    if license_key_index_location:
+        with open(license_key_index_location, 'r') as f:
+            license_key_info = json.load(f)
+    else:
+        curr_dir = dirname(abspath(__file__))
+        data_dir = join(curr_dir, 'data')
+        vendored_license_key_index_location = join(data_dir, 'license_key_index.json')
+        with open(vendored_license_key_index_location, 'r') as f:
+            license_key_info = json.load(f)
+    return license_key_info
+
+
+def build_licensing(license_key_index_location=None):
+    """
+    Return a Licensing object that has been loaded with license keys.
+
+    If `license_key_index_location` is present, then license key information
+    will be loaded from `license_key_index_location`, otherwise license key
+    information will come from a vendored license key index file.
+    """
+    license_key_info = get_license_key_info(license_key_index_location)
+    lics = [
+        {
+            'key': l.get('license_key', ''),
+            'is_exception': l.get('is_exception', ''),
+        } for l in license_key_info
+    ]
+    syms = [LicenseSymbol(**l) for l in lics]
+    return Licensing(syms)
```

Comment:
May be these two lines could be factored out as a function for use here and below?

Reply:
Done.

Reply:
I've put this in a new function `load_licensing_from_license_index`